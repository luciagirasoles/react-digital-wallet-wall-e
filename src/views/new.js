/** @jsx jsx */
import { jsx } from "@emotion/core";
import React from "react";
import { connect } from "react-redux";
import { navigate } from "@reach/router";
import { Input, Card, Select, Button } from "../components/ui";
import { addTransaction } from "../actions";

export function New({ categories, createTransaction }) {
  const [type, setType] = React.useState("ingresse");
  const [amount, setAmount] = React.useState("");
  const [category, setCategory] = React.useState("");

  function handleTypeChange(event) {
    setType(event.target.value);
  }

  function handleAmountChange(event) {
    setAmount(parseFloat(event.target.value));
  }

  function handleCategoryChange(event) {
    setCategory(parseInt(event.target.value));
  }

  function handleSubmit(event) {
    event.preventDefault();
    createTransaction({
      type,
      amount,
      categoryId: category
    });
    navigate("/");
  }

  const styleLabel = {
    marginRight: "1em",
    minWidth: "115px",
    "@media (max-width: 460px)": {
      display: "block",
      margin: "1rem 0"
    }
  };

  const styleField = {
    margin: "2rem 0",
    display: "flex",
    alignItems: "center",
    "@media (max-width: 460px)": {
      flexDirection: "column",
      alignItems: "initial",
      margin: "1rem 0"
    }
  };

  return (
    <Card
      css={{
        margin: "auto",
        maxWidth: "500px"
      }}
    >
      <form onSubmit={handleSubmit}>
        <div css={styleField}>
          <label css={styleLabel}>Type: </label>
          <label>
            <input
              type="radio"
              name="ingresse"
              value="ingresse"
              checked={type === "ingresse"}
              onChange={handleTypeChange}
            />
            Ingresse
          </label>
          <label>
            <input
              type="radio"
              name="withdraw"
              value="withdraw"
              checked={type === "withdraw"}
              onChange={handleTypeChange}
            />
            Withdraw
          </label>
        </div>
        <div css={styleField}>
          <label htmlFor="amount" css={styleLabel}>
            Amout:{" "}
          </label>
          <Input
            aria-label="Write the amount"
            placeholder="1.00"
            type="number"
            id="amount"
            name="amount"
            required
            value={amount}
            onChange={handleAmountChange}
          />
        </div>
        <div css={styleField}>
          <label htmlFor="filter" css={styleLabel}>
            Select Category:
          </label>
          <Select
            styles={{
              container: {
                flex: "1",
                "@media (max-width: 460px)": {
                  width: "100%"
                }
              }
            }}
            id="filter"
            name="filter"
            aria-label="Select a filter"
            onChange={handleCategoryChange}
            value={category}
            required
          >
            <option value="">-- Select a category --</option>
            {categories
              .filter(categorie => categorie.type === type)
              .map(categorie => (
                <option value={categorie.id} key={categorie.id}>
                  {categorie.name}
                </option>
              ))}
          </Select>
        </div>
        <div
          css={{
            margin: "2rem 0"
          }}
        >
          <Button type="submit">Save</Button>
        </div>
      </form>
    </Card>
  );
}

function mapState(state) {
  return {
    categories: Object.values(state.categories),
    data: "space"
  };
}

function mapDispatch(dispatch) {
  return {
    createTransaction(transaction) {
      return dispatch(addTransaction(transaction));
    }
  };
}

export default connect(
  mapState,
  mapDispatch
)(New);
