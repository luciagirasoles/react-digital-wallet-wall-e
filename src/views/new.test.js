import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import { New } from "./new";

test("New component", async () => {
  const categories = [
    {
      id: 1,
      type: "ingresse",
      name: "Award",
      image: "/images/Award.png"
    },
    {
      id: 2,
      type: "ingresse",
      name: "Space",
      image: "/images/Space.png"
    }
  ];
  const createTransaction = jest.fn();

  const { getByLabelText, asFragment } = render(
    <New categories={categories} createTransaction={createTransaction} />
  );

  expect(asFragment()).toMatchSnapshot();

  const amountInput = await waitForElement(() =>
    getByLabelText("Write the amount")
  );

  fireEvent.change(amountInput, { target: { value: 20.0 } });

  const typeFilter = getByLabelText("Select a category");
  fireEvent.change(typeFilter, { target: { value: "Space" } });

  fireEvent.submit(amountInput);

  expect(createTransaction).toHaveBeenCalledWith({
    type: "ingresse",
    amount: 20.0,
    categoryId: 2
  });
});
