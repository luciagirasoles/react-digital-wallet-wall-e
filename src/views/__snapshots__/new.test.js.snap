// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New component 1`] = `
<DocumentFragment>
  .emotion-10 {
  background: white;
  border-radius: .5em;
  box-shadow: 0px 2px 5px 0px rgba(0,0,0,.12);
  box-sizing: border-box;
  padding: 2rem;
  width: 100%;
  margin: auto;
  max-width: 500px;
}

.emotion-1 {
  margin: 2rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (max-width:460px) {
  .emotion-1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: initial;
    -webkit-box-align: initial;
    -ms-flex-align: initial;
    align-items: initial;
    margin: 1rem 0;
  }
}

.emotion-0 {
  margin-right: 1em;
  min-width: 115px;
}

@media (max-width:460px) {
  .emotion-0 {
    display: block;
    margin: 1rem 0;
  }
}

.emotion-3 {
  background: none;
  border: 1px solid #eaeaea;
  border-radius: .25rem;
  box-sizing: border-box;
  display: block;
  font-size: 1rem;
  padding: .5rem;
  width: 100%;
}

.emotion-3:focus {
  outline: none;
  border-color: #00c852;
}

.emotion-6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

@media (max-width:460px) {
  .emotion-6 {
    width: 100%;
  }
}

.emotion-9 {
  margin: 2rem 0;
}

.emotion-8 {
  background-color: #00c852;
  border: 1px solid #00c852;
  border-radius: .25rem;
  color: white;
  cursor: pointer;
  font-size: .8rem;
  padding: .75rem 0;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
  text-align: center;
  text-transform: uppercase;
  width: 100%;
}

.emotion-8:hover {
  background-color: white;
  color: #00c852;
}

<div
    class="emotion-10"
  >
    <form>
      <div>
        <h4
          id="demo"
        >
          Basic, Fixed List Combobox
        </h4>
      </div>
      <div
        class="emotion-1"
      >
        <label
          class="emotion-0"
        >
          Type: 
        </label>
        <label>
          <input
            checked=""
            name="ingresse"
            type="radio"
            value="ingresse"
          />
          Ingresse
        </label>
        <label>
          <input
            name="withdraw"
            type="radio"
            value="withdraw"
          />
          Withdraw
        </label>
      </div>
      <div
        class="emotion-1"
      >
        <label
          class="emotion-0"
          for="amount"
        >
          Amout: 
        </label>
        <input
          aria-label="Write the amount"
          class="emotion-3"
          id="amount"
          name="amount"
          placeholder="1.00"
          required=""
          type="number"
          value=""
        />
      </div>
      <div
        class="emotion-1"
      >
        <label
          class="emotion-0"
          for="filter"
        >
          Select Category:
        </label>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="listbox:1"
          class="emotion-6"
          data-reach-combobox=""
          role="combobox"
        >
          <input
            aria-autocomplete="both"
            aria-label="Select a category"
            autocomplete="off"
            data-reach-combobox-input=""
            id="listbox:1"
            placeholder="Select a category"
            value=""
          />
        </div>
      </div>
      <div
        class="emotion-9"
      >
        <button
          class="emotion-8"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
